third_party/amd/lib/TritonAMDGPUTransforms/Utility.cpp

```diff
@@ -2,24 +2,44 @@
 
 #include "mlir/Dialect/SCF/IR/SCF.h"
 
-namespace findMin {
-int findMinInBlock(Block &block,
-                   const std::function<int(Operation *)> &countFunc);
+namespace deduceMin {
+int deduceMinCountInBlock(Block &block,
+                          const std::function<int(Operation *)> &countFunc);
 
-// Returns the minimum found when accumulating countFunc(op) for all paths
-// between op1 and op2
-int findMinBetweenOps(Operation *op1, Operation *op2,
-                      const std::function<int(Operation *)> &countFunc) {
-  assert(op1 && op2);
+std::optional<int> getIntIfConstant(Value v) {
+  if (auto constantOp = v.getDefiningOp<arith::ConstantOp>()) {
+    if (auto attr = dyn_cast<IntegerAttr>(constantOp.getValue())) {
+      return attr.getInt();
+    }
+  }
+  return std::nullopt;
+}
+
+// Returns the minimum found when accumulating countFunc(op) between begin and
+// end (inclusive)
+int deduceMinCountBetweeOps(Operation *beginOp, Operation *endOp,
+                            const std::function<int(Operation *)> &countFunc) {
+  assert(beginOp && endOp);
+  assert(beginOp == endOp || beginOp->isBeforeInBlock(endOp));
   int count = 0;
-  for (auto op = op1; op != op2; op = op->getNextNode()) {
+  for (auto op = beginOp; op != endOp; op = op->getNextNode()) {
     if (auto ifOp = llvm::dyn_cast<scf::IfOp>(op)) {
       assert(!ifOp.getThenRegion().empty() && !ifOp.getElseRegion().empty());
-      auto minThen = findMinInBlock(ifOp.getThenRegion().front(), countFunc);
-      auto minElse = findMinInBlock(ifOp.getElseRegion().front(), countFunc);
+      auto minThen =
+          deduceMinCountInBlock(ifOp.getThenRegion().front(), countFunc);
+      auto minElse =
+          deduceMinCountInBlock(ifOp.getElseRegion().front(), countFunc);
       count += std::min(minThen, minElse);
     } else if (auto forOp = llvm::dyn_cast<scf::ForOp>(op)) {
-      count += findMinInBlock(*forOp.getBody(), countFunc);
+      auto upperBound = getIntIfConstant(forOp.getUpperBound());
```

Comment:
You can use [`constantTripCount`](https://mlir.llvm.org/doxygen/namespacemlir.html#a694b6ad2c919c4bcef792aca11deda42) from mlir to avoid manually repeat the logic here.